File: test-suite/golang/security/header_injection_buggy.go
package security

import "net/http"

type headerContext interface {
	Query(string) string
	Header(string, string)
	Set(string, string)
}

func queryValueInHeader(w http.ResponseWriter, r *http.Request) {
	displayName := r.URL.Query().Get("name")
	w.Header().Set("X-Display-Name", displayName)
}

func formValueInContentDisposition(w http.ResponseWriter, r *http.Request) {
	filename := r.FormValue("filename")
	w.Header().Set("Content-Disposition", "attachment; filename="+filename)
}

func requestHeaderReflected(w http.ResponseWriter, r *http.Request) {
	traceID := r.Header.Get("X-Trace-ID")
	w.Header().Add("X-Upstream-Trace", traceID)
}

func frameworkHeaderFromQuery(c headerContext) {
	c.Header("X-Return-Reason", c.Query("reason"))
	c.Set("X-Mode", c.Query("mode"))
}

File: test-suite/csharp/security/HeaderInjectionBuggy.cs
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Primitives;

public sealed class HeaderInjectionBuggy : Controller
{
    public void QueryValueInHeader(HttpContext context)
    {
        var displayName = context.Request.Query["name"];
        context.Response.Headers["X-Display-Name"] = displayName!;
    }

    public void RequestFilenameInDisposition(HttpContext context)
    {
        var filename = context.Request.Form["filename"];
        context.Response.Headers.Append("Content-Disposition", "attachment; filename=" + filename);
    }

    public void RequestHeaderReflected(HttpContext context)
    {
        context.Request.Headers.TryGetValue("X-Trace-ID", out StringValues traceId);
        context.Response.Headers.Add("X-Upstream-Trace", traceId.ToString());
    }

    public void AnnotatedQueryParameter([FromQuery] string reason)
    {
        Response.Headers.Set("X-Return-Reason", reason);
    }

    public void TypedContentDisposition(HttpContext context)
    {
        var attachment = context.Request.Query["attachment"].ToString();
        context.Response.GetTypedHeaders().ContentDisposition =
            new System.Net.Http.Headers.ContentDispositionHeaderValue("attachment")
            {
                FileName = attachment
            };
    }
}
